aten/src/ATen/core/type.cpp

```diff
@@ -655,4 +655,90 @@ void checkNoAny(const Type& base, const char* what, const std::string& attrname,
       "' but it contains an Any type. Any types cannot be members of modules, classes, or named tuples.");
 }
 
+namespace {
+bool isSpecialChar(char a) {
+  return a == '[' || a == ',' || a == ']';
+}
+
+class TypeParser {
+ public:
+  explicit TypeParser(const std::string& pythonStr)
+    : pythonStr_(pythonStr), start_(0) {}
+    
```

Comment:
To be consistent with other parsers in PyTorch, this class should have a `next()` method which returns a `std::string` of the next token, and advances to the next state.

Reply:
next() and cur() are added.